### PR TITLE
Create a `UIViewModel`

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

Migrating the ideas from [DropInViewModel](https://github.com/mapbox/mapbox-navigation-android/blob/34402b6c56fcd3b8c2965a4375737989d9501f9d/libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/DropInViewModel.kt#L16) and making them a lifecycle aware `MapboxNavigationObserver`.

`UIViewModel` contains `State` and processes `Action`. Migrated a single example with @abhishek1508 so far. This should give us the complete set of tools we need to build drop in ui.

- *`UICoordinator`  : responsible for assigning a layout a `UIBinder`
- `UIBinder` : inflates views, handles transitions between layouts, and attaches `MapboxNavigationObserver`
- *`UIViewModel` : Same responsibility as a Android ViewModel. Holds Flows, LiveData, is able to update storage, passes data between components, ..
- *`UIComponent` : uses state to update views, sends view events to `UIViewModel`

*`UICoordinator`, `UIViewModel`, and `UIComponent`; are lifecycle aware. They can be used by `LifecycleOwner` because they implement `MapboxNavigationObserver`. They can also create recursive structures. For example, `UICoordinator` can be used inside a `UIBinder`. `UIViewModel` and `UIComponent` can be used without a `UICoordinator`.

### More details

`UIViewModel` is similar to `UIComponent`. The reason we have discussed, creating a separate thing, is because many `UIComponents` are not for controlling user interfaces. They are for controlling **state** and *actions**. The name ViewModel is arguable, listing some of the other ideas discussed:

- [x] UIViewModel (not an actual ViewModel, call it view model)
- [ ] UIBehavior
- [ ] StateManager
- [ ] Controller
- [ ] Manager

The `UIViewModel` also represents, something very close to a `ViewModel`. In fact, that is where they are being defined, inside the `DropInNavigationViewModel`. The reason it is not a `ViewModel`, is because we want to have an instance state that can be shared with Android Auto. For example, an `AudioGuidanceViewModel` that implements `MapboxNavigationObserver` is easily shared with `CarAppService`.

### CODE REVIEW

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt [hunk 1]
```diff
@@ -158,17 +157,19 @@ class DropInNavigationView @JvmOverloads constructor(
     private fun initMapLongClickListener() {
         binding.mapView.gestures.addOnMapLongClickListener { clickPoint ->
             ifNonNull(MapboxNavigationApp.current()) { mapboxNavigation ->
```
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt:159]: Could move this inside the behavior. But it requires location permissions

Comment (cafesilencio) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt:159]: It was moved in this PR: https://github.com/mapbox/mapbox-navigation-android/pull/5496

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UICoordinator.kt [hunk 2]
```diff
@@ -19,13 +22,15 @@ import kotlinx.coroutines.launch
...
     private val viewGroup: T
-) : UIComponent() {
+) : MapboxNavigationObserver {
+
+    lateinit var coroutineScope: CoroutineScope
 
     @CallSuper
     override fun onAttached(mapboxNavigation: MapboxNavigation) {
         var attachedObserver: MapboxNavigationObserver? = null
+        coroutineScope = MainScope()
```
Comment (cafesilencio) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UICoordinator.kt:32]: For consistency I suggest using the InternalJobControlFactory to get a CoroutineScope.

**Review by cafesilencio — APPROVED** [review]

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UIViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import kotlinx.coroutines.launch
+
+/**
+ * Implement your version of a behavior. Behaviors do not reference android ui elements directly,
+ * this is because their lifecycle will survive beyond a view or activity.
+ *
+ * Behaviors have a lifecycle, contain state, and process actions. [UIComponent] will respond
+ */
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+abstract class UIViewModel<State, Action>(initialState: State) : MapboxNavigationObserver {
```
Comment (tomaszrybakiewicz) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UIViewModel.kt:23]: This looks more like a [Redux Reducer](https://redux.js.org/tutorials/essentials/part-1-overview-concepts#redux-application-data-flow) than a [ViewModel](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93viewmodel).

Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UIViewModel.kt:23]: Yes you are definitely correct! `UIReducer`? Right now it's stolen from the existing view model.

https://github.com/mapbox/mapbox-navigation-android/blob/c8f6f3b73ec1bba3242e8b31c7dd1c0e2c801256/libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/DropInViewModel.kt#L16

Let's bring this up in the meeting tomorrow as we figure out names for the things we want. Requirements are:

- Holds state between orientation changes
- Is lifecycle aware (onAttach onDetach)

**Pull request merged**